aws-lc-rs/src/cipher.rs

```diff
@@ -3,158 +3,1270 @@
             return Err(Unspecified);
         }
 
-        unsafe {
-            let mut aes_key = MaybeUninit::<AES_KEY>::uninit();
-            #[allow(clippy::cast_possible_truncation)]
-            if 0 != AES_set_encrypt_key(
-                key_bytes.as_ptr(),
-                (key_bytes.len() * 8) as c_uint,
-                aes_key.as_mut_ptr(),
-            ) {
-                return Err(Unspecified);
+        Ok(PaddedBlockDecryptingKey {
+            key,
+            mode,
+            padding,
+            context,
+        })
+    }
+
+    /// Returns the cipher algorithm.
+    #[must_use]
+    pub fn algorithm(&self) -> &Algorithm {
+        self.key.algorithm()
+    }
+
+    /// Returns the cipher operating mode.
+    #[must_use]
+    pub fn mode(&self) -> OperatingMode {
+        self.mode
+    }
+
+    /// Returns the cipher padding strategy.
+    #[must_use]
+    pub fn padding(&self) -> PaddingStrategy {
+        self.padding
+    }
+
+    /// Decrypts and unpads data provided in `in_out` in-place.
+    /// Returns a references to the decrypted data.
+    ///
+    /// # Errors
+    /// * [`Unspecified`]: Returned if decryption fails.
+    ///
+    pub fn decrypt(self, in_out: &mut [u8]) -> Result<&mut [u8], Unspecified> {
+        let block_len = self.algorithm().block_len();
+        let padding = self.padding;
+        let mut in_out = self.into_decrypting_key()?.decrypt(in_out)?;
+        in_out = padding.remove_padding(block_len, in_out)?;
+        Ok(in_out)
+    }
+
+    fn into_decrypting_key(self) -> Result<DecryptingKey, Unspecified> {
+        DecryptingKey::new(self.key, self.mode, self.context)
+    }
+}
+
+impl Debug for PaddedBlockDecryptingKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PaddedBlockDecryptingKey")
+            .field("key", &self.key)
+            .field("mode", &self.mode)
+            .field("padding", &self.padding)
+            .field("context", &self.context)
+            .finish()
+    }
+}
+
+/// A cipher encryption key that does not perform block padding.
+pub struct EncryptingKey {
+    key: UnboundCipherKey,
+    mode: OperatingMode,
+    context: CipherContext,
+}
+
+impl EncryptingKey {
+    /// Constructs an `EncryptingKey` operating in counter (CTR) mode using the provided key.
+    ///
+    /// # Errors
+    /// * [`Unspecified`]: Returned if there is an error construct the `EncryptingKey`.
+    ///
+    pub fn ctr(key: UnboundCipherKey) -> Result<EncryptingKey, Unspecified> {
+        EncryptingKey::new(key, OperatingMode::CTR, None)
+    }
+
+    /// Constructs an `EncryptingKey` operating in counter (CTR) mode using the provided key.
+    /// The users provided context will be used for the CTR mode initalization-vector.
+    ///
+    /// # Errors
+    /// * [`Unspecified`]: Returned if there is an error creating the `EncryptingKey`.
+    ///
+    pub fn less_safe_ctr(
```

Comment:
Still pondering on the name. It's not inherently less safe than the other function call....